packages/generator/ni_measurement_plugin_sdk_generator/client/_support.py

```diff
@@ -209,6 +212,18 @@ def get_configuration_and_output_metadata_by_index(
             default_value = default_value.value
         elif isinstance(default_value, list) and any(isinstance(e, Enum) for e in default_value):
             default_value = [e.value for e in default_value]
+        elif isinstance(default_value, str):
+            default_value = repr(default_value)
+        elif configuration_metadata[id].message_type == "ni.protobuf.types.Double2DArray":
```

Comment:
Can you give an example of what 'default_value' is when it comes in here and what it looks like after you reconstruct it?

Reply:
I think this approach is ok. I'd like to hear what @bkeryan or @mshafer-NI think.

Reply:
Default Value for Double2DArray before reconstruct -
rows: 2
columns: 3
data: 1
data: 2
data: 3
data: 4
data: 5
data: 6

After Reconstruct -
'Double2DArray(rows=2, columns=3, data=[1, 2, 3, 4, 5, 6])'

Reply:
Now, the error (without reconstruct) is thrown because the generated template has Double2DArray value as `default_value=rows: 2\ncolumns: 3\ndata: 1\ndata: 2\ndata: 3\ndata: 4\ndata: 5\ndata: 6\n,\r\n`

and it is passed to this [method](https://github.com/ni/measurement-plugin-python/blob/7ea52e5a2b45da2af95e3f66e0563a41239ccfdd/packages/generator/ni_measurement_plugin_sdk_generator/client/__init__.py#L50)

Reply:
```
AssertionError: assert 1 == 0
E        +  where 1 = <Result InvalidInput('Cannot parse: 242:34:                 default_value=rows: 2')>.exit_code

tests\acceptance\test_client_generator.py:44: AssertionError
```

Reply:
@bkeryan If I understand correctly, your point here is that `default_value` should contain the python data type (`Double2DArray` in this case) and any formatting should be done in the `_format_default_value` method defined in the template file.
`_format_default_value` should be checking if `default_value` is an instance of Double2DArray and do the necessary steps to convert it to the `'Double2DArray(rows=2, columns=3, data=[1, 2, 3, 4, 5, 6])'` string. 

That makes sense. 

If the above understanding is correct then. one concern is that currently the `default_value` is added to the generated file not only through the `_format_default_value` method. [Here](https://github.com/ni/measurement-plugin-python/blob/7ea52e5a2b45da2af95e3f66e0563a41239ccfdd/packages/generator/ni_measurement_plugin_sdk_generator/client/_support.py#L263C81-L263C94) the default value is being used directly and the reason that things work fine is because [this piece of code](https://github.com/ni/measurement-plugin-python/blob/7ea52e5a2b45da2af95e3f66e0563a41239ccfdd/packages/generator/ni_measurement_plugin_sdk_generator/client/_support.py#L229C1-L230C48) incorporates the [special string logic in _format_default_value](https://github.com/ni/measurement-plugin-python/blob/7ea52e5a2b45da2af95e3f66e0563a41239ccfdd/packages/generator/ni_measurement_plugin_sdk_generator/client/templates/measurement_plugin_client.py.mako#L10C1-L11C31).
Is that intentional? 
If yes then, whatever `Double2DArray` conversion logic that is added into `_format_default_value` will need to be added here as well, or we would need this code to be passed through the `_format_default_value` workflow.

Reply:
@clintverghese It may look like this is duplicated code, but `get_configuration_parameters_with_type_and_default_values` and `_format_default_value` are doing slightly different things.

`get_configuration_parameters_with_type_and_default_values`  generates Python code for the client API class. In some cases we use more Pythonic data types in the client API. For example, path parameters are `pathlib.Path` in the client API and `str` in protobuf. 

`_format_default_value` is used when generating Python code for the service's metadata structs. This tends to use the protobuf types or basic Python types, not stuff like `pathlib.Path`.

I think it would be fine to single-source `_format_default_value` by moving it into `_support.py` so that `get_configuration_parameters_with_type_and_default_values` can call it for non-path cases. Also, it would be helpful to update the docstrings to clarify this difference.

Reply:
Moved the `_format_default_value` method to `_support.py` and added logic for Double2DArray. Now, we are calling this method from `get_configuration_parameters_with_type_and_default_values` for string and Double2dArray parameter types in `_support.py` and for the default value conversions in the template. Also, updated Docstring for both the methods.